Refactor optbib to using a git submodule (and bugfix)

DE0CH opened the pull request:
Hi! I think it's better to use a git submodule as it is well designed to handle a a git repo inside another git repo.

This PR also fixes three bugs in `setup.sh`:
1. Misspelling of `install.packages`
2. Specifying a mirror for R because otherwise R will ask for input, causing the script to fail.
3. Add `-y` for installing `bibtex2html` to make the script not ask for user input.

Thanks.

DE0CH commented:
Thanks for the feedback. There are a few changes.

1. The part that uses apt in `scripts/setup.sh` has been moved to `scripts/setup-deb.sh`. The `setup-deb.sh` script has been tested to work on Ubuntu and Debian.
2. I removed the installation for `bibtex2html` because nothing seems to need to use it.
3. I installed the dependencies for generating the the vignettes using `apt` instead of `tlmgr` because the latter is a bit broken because of server problems with TeX Live. 
4. Added a `Dockerfile` to make it easier for folks using windows, mac and other linux distros to build the package without worrying about dependencies and environment. 

I have tested the changes on the LTS releases of Ubuntu and stable releases of Debian. All the tests in `make check` passed on Ubuntu Focal (20.4) and Debian Bullseye (11). There is some bug with the test suite for Ubuntu Bionic (18.04) (see Issue #26), but it does install and build. For older versions I ran into all kinds of issue with installing `devtools`, and `testthat`. At this point I think R probably dropped support for those versions so there's nothing I can do.

MLopez-Ibanez commented on Makefile:
```diff
@@ -134,13 +136,12 @@ clean:
 #
 # It is ok to fail if something is missing or needs to be done and give a nice error. For example, if optbib is missing.
-vignettes: vignettes/$(PACKAGE)-package.Rnw vignettes/section/irace-options.tex
+vignettes: vignettes/$(PACKAGE)-package.Rnw vignettes/section/irace-options.tex setup
```
Sorry to be picky, but vignettes should not depend on setup. Nothing should depend on setup. Setup should be a manual step done once by a developer.

MLopez-Ibanez commented on Makefile:
```diff
@@ -134,13 +136,12 @@ clean:
+vignettes: vignettes/$(PACKAGE)-package.Rnw vignettes/section/irace-options.tex setup
 # FIXME: How to display the output of the latex and bibtex commands with R CMD?
 # FIXME: How to halt on warning?
-	@test -d $(PACKAGEDIR)/vignettes/optbib || (echo "ERROR: vignettes/optbib not found. You need to symlink or checkout https://github.com/iridia-ulb/references." && exit 1)
```
The test should remain and if it fails, it should try "git submodule init && git submodule update"

MLopez-Ibanez commented on scripts/setup.sh:
```diff
@@ -1,15 +1,10 @@
 EOF
-# How to make this more general for any Linux/MacOS ?
-# Maybe use bibtool -x instead?
-command -V aux2bib || (echo "aux2bib not found, installing..."; sudo apt install bibtex2html)
```
aux2bib is used by Makefile, so we need to check whether it is installed and if not, install the package bibtex2html

DE0CH replied:
I moved into the installation to inside the function.

MLopez-Ibanez commented on scripts/setup.sh:
```diff
@@ -1,15 +1,10 @@
-# Maybe use bibtool -x instead?
-command -V aux2bib || (echo "aux2bib not found, installing..."; sudo apt install bibtex2html)
-
-test -e ./vignettes/optbib || (echo "vignettes/optbib not found, cloning from github..."; git clone https://github.com/iridia-ulb/references.git vignettes/optbib)
```
This script should init the submodule if vignettes/optbib does not exist.

MLopez-Ibanez commented:
>     1. The part that uses apt in `scripts/setup.sh` has been moved to `scripts/setup-deb.sh`. The `setup-deb.sh` script has been tested to work on Ubuntu and Debian.

OK, but the script should be called by `setup.sh`. The purpose of the script is to simplify the setup for a new developer. If one needs to remember extra steps, the purpose is defeated.

>     2. I removed the installation for `bibtex2html` because nothing seems to need to use it.

It provides aux2bib, which is needed.

>     3. I installed the dependencies for generating the the vignettes using `apt` instead of `tlmgr` because the latter is a bit broken because of server problems with TeX Live.

I really prefer tlmgr because it install only the packages that are needed and it also allows a local installation. I believe it also works in MacOS. But OK, let's leave it as you propose for now and let's see how things evolve.
 
>     4. Added a `Dockerfile` to make it easier for folks using windows, mac and other linux distros to build the package without worrying about dependencies and environment.

Great!

> I have tested the changes on the LTS releases of Ubuntu and stable releases of Debian. All the tests in `make check` passed on Ubuntu Focal (20.4) and Debian Bullseye (11). There is some bug with the test suite for Ubuntu Bionic (18.04) (see Issue #26), but it does install and build. For older versions I ran into all kinds of issue with installing `devtools`, and `testthat`. At this point I think R probably dropped support for those versions so there's nothing I can do.

Thanks! Let's not worry too much about 18.04 or older. I left some comments above. 

By the way, I mentioned your name in the NEWS.md file: https://mlopez-ibanez.github.io/irace/news/index.html#new-features-and-improvements-3-5 
I hope this is OK, otherwise let me know and I will delete your name.

DE0CH commented on Makefile:
```diff
@@ -137,10 +139,11 @@ clean:
 # FIXME: How to display the output of the latex and bibtex commands with R CMD?
 # FIXME: How to halt on warning?
-	@test -d $(PACKAGEDIR)/vignettes/optbib || (echo "ERROR: vignettes/optbib not found. You need to symlink or checkout https://github.com/iridia-ulb/references." && exit 1)
+	git submodule init 
+	git submodule update
```
I didn't check if the folder exists because I want to leave this to git submodule, which would probably check the status of the folder much better than any script I can write.

MLopez-Ibanez replied:
Will `git submodule init` not give an error if optbib already exists? I think doing that work if optbib already exists is not very useful.

DE0CH replied:
No I checked that git will just return 0 without doing anything if it is initialised.

DE0CH replied:
Actually I think you're right. I have added a test before calling git submodule update. I think at the very least it won't automatically change the content of the submodule if the users has modified it.

DE0CH commented:
Thanks for the feedback, I just made some edits
> OK, but the script should be called by `setup.sh`. The purpose of the script is to simplify the setup for a new developer. If one needs to remember extra steps, the purpose is defeated.

I just moved the apt setup inside `setup.sh` as a function. I didn't figure out how to setup the system properly for other linux distros that don't use apt like archlinux, alpine, etc. But I guess just having it work on Debian/Ubuntu should be good enough. 
 
> It provides aux2bib, which is needed.

Sorry, didn't know that. I just added it back. 

> I really prefer tlmgr because it install only the packages that are needed and it also allows a local installation. I believe it also works in MacOS. But OK, let's leave it as you propose for now and let's see how things evolve.

I see. I was not really familiar with `tlmgr` so perhaps it's a bit foreign to me. But `apt` seems to work well at least in a docker container. Installing `tlmgr` is a bit finicky because the version installed with `apt` would not work, and installiing it from the installer downloaded from ctan requires some extra scripts to configure it properly. The more scripts I have to write the more buggy it could get (and it might just stop working in a year or two after they update the version) so I just chose to install the fonts with `apt`. 

> Thanks! Let's not worry too much about 18.04 or older. I left some comments above.

Sure. 

> By the way, I mentioned your name in the NEWS.md file: https://mlopez-ibanez.github.io/irace/news/index.html#new-features-and-improvements-3-5 I hope this is OK, otherwise let me know and I will delete your name.

Yeah that's absolutely find. Thanks for giving me credit!

MLopez-Ibanez commented on scripts/setup.sh:
```diff
@@ -1,15 +1,38 @@
+    apt-get install -y texlive-fonts-extra
+    apt-get install -y texlive-latex-extra
+    apt-get install -y texlive-science 
+    apt-get install -y aux2bib
```
Have you tested this? I think there is no such package aux2bib. The package name is `bibtex2html`.
But also, if the executable already exists, because the user installed it in some other way, it is useless to install this package. So please do:

command -V aux2bib || apt-get install -y bibtex2html

DE0CH replied:
Sorry about this. I did test the installation, the tests passed but I didn't see the error message that aux2bib was not found. Will fix this.

MLopez-Ibanez commented on Makefile:
```diff
@@ -45,6 +45,8 @@ help:
 setup:
 	./scripts/setup.sh
+	git submodule init
+	git submodule update
```
The git commands should also be moved within setup.sh.

DE0CH replied:
I put them here because I wanted to make sure that `setup.sh` can be run from anywhere even without a full view of the git repo. One of the benefits is in the docker installation. If I can just copy and run `setup.sh` before I copy the entire repo into the container, the installation for all the apt packages will be cached so the user would not need to waste time installing the packages after they edit the repo and rebuild the package with docker.

DE0CH replied:
Actually I think you're right. I have added a test before calling git submodule update. I think at the very least it won't automatically change the content of the submodule if the users has modified it.
